materialize-postgres-rc/driver.go

```diff
@@ -0,0 +1,383 @@
+	return nil
+}
+
+// ToURI converts the Config to a DSN string.
+func (c *config) ToURI() string {
+	var address = c.Address
+	var uri = url.URL{
+		Scheme: "postgres",
+		Host:   address,
+		User:   url.UserPassword(c.User, c.Password),
+	}
+	if c.Database != "" {
+		uri.Path = "/" + c.Database
+	}
+	// Compatibility with pgBouncer. See:
+	// https://github.com/jackc/pgx/issues/650
+	uri.RawQuery = "statement_cache_mode=describe"
+	return uri.String()
+}
+
+type tableConfig struct {
+	Schema string `json:"schema,omitempty" jsonschema:"title=Alternative Schema,description=Alternative schema for this table (optional)"`
+	Table  string `json:"table" jsonschema:"title=Table,description=Name of the database table"`
+	Delta  bool   `json:"delta_updates,omitempty" jsonschema:"default=false,title=Delta Update,description=Should updates to this table be done via delta updates. Defaults is false."`
+}
+
+func newTableConfig(ep *sql.Endpoint) sql.Resource {
+	return &tableConfig{
+		// Default to an explicit endpoint configuration schema, if set.
+		// This will be over-written by a present `schema` property within `raw`.
+		Schema: ep.Config.(*config).Schema,
+	}
+}
+
+// Validate the resource configuration.
+func (r tableConfig) Validate() error {
+	if r.Table == "" {
+		return fmt.Errorf("missing table")
+	}
+	return nil
+}
+
+func (c tableConfig) Path() sql.TablePath {
+	if c.Schema != "" {
+		return []string{c.Schema, c.Table}
+	}
+	return []string{c.Table}
+}
+
+func (c tableConfig) DeltaUpdates() bool {
+	return c.Delta
+}
+
+func newPostgresDriver() pm.DriverServer {
+	return &sql.Driver{
+		DocumentationURL: "https://go.estuary.dev/materialize-postgresql",
+		EndpointSpecType: new(config),
+		ResourceSpecType: new(tableConfig),
+		NewEndpoint: func(ctx context.Context, raw json.RawMessage) (*sql.Endpoint, error) {
+			var cfg = new(config)
+			if err := pf.UnmarshalStrict(raw, cfg); err != nil {
+				return nil, fmt.Errorf("parsing endpoint configuration: %w", err)
+			}
+
+			log.WithFields(log.Fields{
+				"database": cfg.Database,
+				"address":  cfg.Address,
+				"user":     cfg.User,
+			}).Info("opening database")
+
+			var metaBase sql.TablePath
+			if cfg.Schema != "" {
+				metaBase = append(metaBase, cfg.Schema)
+			}
+			var metaSpecs, metaCheckpoints = sql.MetaTables(metaBase)
+
+			return &sql.Endpoint{
+				Config:              cfg,
+				Dialect:             pgDialect,
+				MetaSpecs:           metaSpecs,
+				MetaCheckpoints:     &metaCheckpoints,
+				Client:              client{uri: cfg.ToURI()},
+				CreateTableTemplate: tplCreateTargetTable,
+				NewResource:         newTableConfig,
+				NewTransactor:       newTransactor,
+			}, nil
+		},
+	}
+}
+
+// client implements the sql.Client interface.
+type client struct {
+	uri string
+}
+
+func (c client) FetchSpecAndVersion(ctx context.Context, specs sql.Table, materialization pf.Materialization) (specB64, version string, err error) {
+	err = c.withDB(func(db *stdsql.DB) error {
```

Comment:
I did it this way because, in practice, each invocation of the connector calls approximately O(1) of these functions touching the DB, and we're not all that sensitive to latency since it happens so rarely, and it simplified the `client` implementation (for example, making it obvious when a sql.DB can be torn down). It also felt disingenuous to have a DB pool which isn't actually used by the meat & potatoes of the Transactions RPC.

pgBouncer would have been fine with the previous approach, too, that wasn't a concern here.